File: server/dto/error_dto.py
from dataclasses import dataclass

from dto.base_dto import BaseDto
from exceptions.custom_exception import CustomException


@dataclass
class ErrorDto(BaseDto):
    error: str
    displayToUser: bool
    socketConnectionKiller: bool

    @classmethod
    def from_exception(cls, ex: Exception):
        dict_repr = {"error": str(ex)}

        if isinstance(ex, CustomException):
            dict_repr["displayToUser"] = True
            dict_repr["socketConnectionKiller"] = ex.socket_connection_killer
        else:
            dict_repr["displayToUser"] = False
            dict_repr["socketConnectionKiller"] = False

        return dict_repr

File: client/src/dto/ErrorDto.ts
export interface ErrorDto {
    error: string;
    displayToUser: boolean;
    socketConnectionKiller: boolean;
}

export function ParseErrorDto(data: unknown) {
    if (isErrorDto(data)) {
        return {
            error: data.error,
            displayToUser: data.displayToUser,
            socketConnectionKiller: data.socketConnectionKiller,
        } as ErrorDto;
    } else {
        return { error: data, displayToUser: false, socketConnectionKiller: false } as ErrorDto;
    }
}

function isErrorDto(data: unknown): data is ErrorDto {
    return typeof data === "object" && data !== null && data !== undefined &&
        "error" in data && "displayToUser" in data && "socketConnectionKiller" in data;
}